### PR TITLE
Add Responses API with Copilot support

### PR DESCRIPTION
**Pull request by karthink**
Implement a new `gptel-openai-responses` backend implementing support for the OpenAI responses API.

Differs from both #1168 in #1218 in that this is a first-class backend (inheriting from `gptel-backend`) instead of a set of methods dispatched on from the `gptel-openai` backend based on the endpoint in use.

The default OpenAI backend will use `gptel-make-openai-responses` instead of `gptel-make-openai`.

Overall, it is cleaner to implement Responses API support a separate first-class backend.

Copilot support for the Responses API differs from #1218 in that dispatch depending on Copilot models occurs in gptel-gh defmethods.  (I'm looking for some way to avoid even this, finding a way to reduce this dispatch to just one invocation, setting the :backend slot in the request INFO plist before any of the defmethods are called.)

Both #1168 and #1218 also contain several subtle (and some glaring) bugs in prompt construction and response parsing that are difficult to track down and fix, requiring rewriting several functions.

@kiennq Could you take a look at the changes to gptel-gh.el (second commit) and let me know if this approach looks okay to you?

### CODE REVIEW

**Comment by karthink** [comment]
Also, I can't test the `gptel-gh` changes because I don't have access to the gpt-5.x models using Copilot.  If you could test it it would be great.

I can confirm that the `gptel-openai-responses` backend works well, although there might be edge cases.

**Comment by bjodah** [comment]
@karthink I tried to evaluate your this branch with my copilot subscription, but I'm probably doing somehing wrong:
from my `init.el`
```elisp
...
(gptel-make-gh-copilot "Copilot")
...
```
`*scratch*`
```elisp
(setq gptel-stream nil)                                                                                                                                                                                                                                                       
nil                                                                                                                                                                                                                                                                           
(setq gptel-log-level 'debug)                                                                                                                                                                                                                                                 
debug                                                                                                                                                                                                                                                                         
(gptel-get-backend "Copilot")                                                                                                                                                                                                                                                 
#s(gptel--gh "Copilot" "api.githubcopilot.com" #[0 "<...some binary data...>" [gptel-backend cl-struct-gptel--gh-tags gptel-track-media gptel-model gptel--gh-auth ("openai-intent" . "conversation-panel") "authorization" "Bearer " plist-get type-of signal              
wrong-type-argument ...] 12] "https" t "/chat/completions" nil (gpt-4.1 gpt-4o gpt-5-mini gpt-5.1 gpt-5.1-codex gpt-5.1-codex-max gpt-5.1-codex-mini gpt-5.2 gpt-5.2-codex claude-haiku-4.5 claude-opus-4.5 claude-opus-4.6 ...)                                              
"https://api.githubcopilot.com/chat/completions" nil nil ...)
```
So I can already tell that this will not work (`/chat/completions`) in the url? I'm trying to make sense of `cl-defun gptel-make-gh-copilot` in `gptel-gh.el` which I'm probably failing to instruct to use the responses-backend.

<details>
<summary>
*gptel-log* after selecting `Copilot:gpt-5.2` in a *gptel* buffer, and asking `captial of france?`
</summary>

```
{
  "gptel": "request headers",
  "timestamp": "2026-03-16 14:19:26"
}
{
  "Content-Type": "application/json",
  "openai-intent": "conversation-panel",
  "authorization": "Bearer REDACTED",
  "x-request-id": "efe7f057-baa9-4b6c-80eb-77fa3b463411",
  "vscode-sessionid": "",
  "vscode-machineid": "REDACTED",
  "copilot-integration-id": "vscode-chat"
}
{
  "gptel": "request body",
  "timestamp": "2026-03-16 14:19:26"
}
{
  "model": "gpt-5.2",
  "input": [
    {
      "role": "user",
      "content": "capital of france?"
    }
  ],
  "store": false,
  "stream": false,
  "instructions": "You are a large language model living in Emacs and a helpful assistant. Respond concisely."
}
{"gptel": "request Curl command", "timestamp": "2026-03-16 14:19:26"}
curl \
--disable \
--location \
--silent \
--compressed \
-XPOST \
-y7200 \
-Y1 \
-D- \
-w\(f18e45a49bf7e4ed5c1667a52ffa8acc\ .\ \%\{size_header\}\) \
-d\{\"model\"\:\"gpt-5.2\"\,\"input\"\:\[\{\"role\"\:\"user\"\,\"content\"\:\"capital\ of\ france\?\"\}\]\,\"store\"\:false\,\"stream\"\:false\,\"instructions\"\:\"You\ are\ a\ large\ language\ model\ living\ in\ Emacs\ and\ a\ helpful\ assistant.\ Respond\ concisely.\"\} \
-HContent-Type\:\ application/json \
-Hopenai-intent\:\ conversation-panel \
-Hauthorization\:\ Bearer\ REDACTED \
-Hx-request-id\:\ efe7f057-baa9-4b6c-80eb-77fa3b463411 \
-Hvscode-sessionid\:\  \
-Hvscode-machineid\:\ REDACTED \
-Hcopilot-integration-id\:\ vscode-chat \
https\://api.githubcopilot.com/chat/completions
{
  "gptel": "response headers",
  "timestamp": "2026-03-16 14:19:27"
}
"HTTP/2 400 \r\ncontent-security-policy: default-src 'none'; sandbox\r\ncontent-type: text/plain; charset=utf-8\r\nstrict-transport-security: max-age=31536000\r\nx-content-type-options: nosniff\r\nx-request-id: efe7f057-baa9-4b6c-80eb-77fa3b463411\r\ndate: Mon, 16 Mar 2026 13:19:27 GMT\r\ncontent-length: 61\r\nx-github-backend: Kubernetes\r\nx-github-request-id: 1F98:2C232E:279031:2B0973:69B8035F\r\n\r"
{
  "gptel": "response body",
  "timestamp": "2026-03-16 14:19:27"
}
{
  "error": {
    "message": "messages must be non-empty",
    "code": ""
  }
}
```

</details>

So it's trying to use a combination of responses and chat completion, note (from above `*gptel-log*`):
```
...
"input": [
    {
      "role": "user",
      "content": "capital of france?"
    }
...
```

Maybe I still have stale .elc files? But I don't thinnk so (no competing gptel* folders in elpa/ just this checkout, did `rm *.elc`, even `package-recompile` on gptel).

**Comment by karthink** [comment]
@bjodah Thanks for testing.  I see the problem with determining the URL, I'll update it soon.

**Comment by karthink** [comment]
@bjodah The URL should be picked based on the `response-api` capability now, please update and test again.

**Comment by bjodah** [comment]
I just tried your latest commit. It seems to be working great, thank you!
<details>
<summary>sample gptel session</summary>

```
Copilot Ready                  [5 tools] [Ignoring media] [Prompt: default] [gpt-5.2]
### capital of france?                                                               
                                                                                     
Paris.                                                                               
                                                                                     
### what's this repo?                                                                
                                                                                     
                                                                                     
``` tool (Glob :pattern "README*" :path "." :depth 4)...                             
                                                                                     
                                                                                     
``` tool (Grep :regex "(README|readme)" :path "." :glob "*.md" :context_lines 2)...  
                                                                                     
                                                                                     
``` tool (Glob :pattern "*" :path "." :depth 2)...                                   
                                                                                     
                                                                                     
``` tool (Read :file_path "/work/gptel.el" :start_line 1 :end_line 60)...            
                                                                                     
                                                                                     
``` tool (Read :file_path "/work/test/README.org" :start_line 1 :end_line 120)...    
                                                                                     
                                                                                     
``` tool (Read :file_path "/work/README.org" :start_line 1 :end_line 120)...         
                                                                                     
                                                                                     
This repo is **gptel**, an **Emacs Lisp package** that provides a simple **LLM chat  
client inside Emacs**.                                                               
                                                                                     
- Lets you interact with LLMs from any buffer, with responses in Org/Markdown.       
- Supports many backends (OpenAI/ChatGPT, Anthropic, Gemini, Groq, Perplexity,       
OpenRouter, Azure, AWS Bedrock, etc.) and local models (Ollama, llama.cpp, GPT4All,  
…).                                                                                  
- Includes features like multi-session chats, tool use/agentic workflows, MCP        
integration, multimodal input (images/docs), and saving/resuming chats.              
- The `test/` directory contains an ERT-based test suite for gptel.                  
                                                                                     
### 
```

</details>

[gptel-log.txt](https://github.com/user-attachments/files/26042932/gptel-log.txt)

**Comment by karthink** [comment]
@bjodah Thanks for testing.

I'm waiting for @kiennq to review my changes and approach since they authored gptel-gh.  I'll merge some version of this PR after that.  In the meantime you can stick to this branch.

**Comment by bjodah** [comment]
@karthink I don't know if this is a known issue, but I get errors in streaming mode when there are tool calls, e.g. if I stand in gptel's repo, and ask "what's this repo?" (with `gptel-agent` tools enabled for: Grep, Read & Glob), I get in `*Messages*` with streaming mode: `Copilot error: (HTTP/2 400) No tool call found for function call output with call_id call_dZmy4DrLLYnPhkN0cVZyxUXH.`
```
Copilot Error: HTTP/2 400    [3 tools] [Ignoring media] [Prompt: default] [gpt-5.2]
### what's this repo?



``` tool (Glob :pattern "*" :path "." :depth 3)...
``` tool (Read :file_path "README.md" :start_line 1 :end_line 200)...
``` tool (Glob :pattern "*.org" :path "." :depth 4)...
``` tool (Glob :pattern "*.md" :path "." :depth 4)...





-UUU:@**- F2  *gptel*        All   L55    (Markdown Fly/en WK GPT ElDoc Wrap) -----
```
I'm attaching logs for both streaming and non-streaming modes if that's helpful / if you're interested. The same query in non-streaming mode finishes without errors.

[gptel-stream.log.txt](https://github.com/user-attachments/files/26086697/gptel-stream.log.txt)
[gptel-non-stream.log.txt](https://github.com/user-attachments/files/26086698/gptel-non-stream.log.txt)

**Comment by karthink** [comment]
@bjodah It was a bug in the streaming parser, now fixed!  Please update and try again.

**Comment by bjodah** [comment]
@karthink that did the trick! thank you.

**Comment by KaoruBB** [comment]
Thanks for this PR!

I tested `feature/openai-responses-api` locally with GitHub Copilot and added the following models to `gptel-gh.el` with the `response-api` capability:

- `gpt-5.3-codex`
- `gpt-5.4`
- `gpt-5.4-mini`

All three worked for me on this branch, and requests were correctly sent to `/v1/responses`.
If you'd like, I can open a small follow-up PR against `feature/openai-responses-api` with these additional models.

**Comment by karthink** [comment]
> All three worked for me on this branch, and requests were correctly sent to `/v1/responses`.
> If you'd like, I can open a small follow-up PR against `feature/openai-responses-api` with these additional models.

@karoubb sure, please do.

**Comment by bjodah** [comment]
@karthink for my own benefit I had gpt-5.4 and opus-4.6 write a chatgpt-codex backend much like the one in #1260, but instead based on this branch. It works well, and the tool calling issue I saw earlier in #1260 is now gone. If you're curious (beware it's written 95%+ by AI) the commit is https://github.com/bjodah/gptel/commit/06b6cfeb3f2bf4f4dbc4dd6a689502db5bec589a. I know that you've asked for a unified oauth framework (which sounds reasonable), ~~but I did not attempt to implement that just yet~~.
EDIT: After some back and forth with multiple LLMs, Gemini 3.1 Pro came up with an "oauth-deduplicated" solution for both gh and chatgpt-codex: https://github.com/bjodah/gptel/pull/5

**Comment by KaoruBB** [comment]
@karthink Thanks!
I've created a PR against the `feature/openai-responses-api` branch: #1302

File: gptel-openai-responses.el [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+                   (plist-put info :tool-use
+                              (cons tool-call (plist-get info :tool-use)))
+                   (plist-put info :partial_json nil)))
+                ;; Reasoning content
+                ("response.reasoning_summary_text.delta"
+                 (when-let* ((delta (plist-get data :delta)))
+                   (plist-put info :reasoning
+                              (concat (plist-get info :reasoning) delta))))
+                ("response.reasoning_summary_text.done"
+                 (plist-put info :reasoning-block t))
```
[the remark this answers is not in the document]
Comment (karthink) [reply gptel-openai-responses.el]: @kerncos Could you generate a log of the response from oss120 or Gemini that contain these blocks?  I can add them to the parser based on that.

**Comment by karthink** [comment]
Merged.  To move OpenAI API users over to the Responses API without causing lots of config churn, `gptel-make-openai` will now create a Completions API or Responses API backend depending on the host (`"api.openai.com"`).

`gptel-make-openai-responses` is available for explicit construction, of course.

Copilot should be working for all models now, but there are edge cases that need to be resolved.  In particular, if you change `gptel-model` while a Copilot request is ongoing in a buffer (tool call turns etc), the request may fail.  Resolving this will take some effort.

-----

@kiennq when you have the time, please review the changes to `gptel-gh.el` and let me know if you have a better idea for how to do the dispatch between the APIs.  The dispatch needs to happen _somewhere_, I picked `gptel-gh` as the place since that's the backend that needs to use _both_ the Completions and Responses API in a single backend.
